services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java

```diff
@@ -127,6 +128,12 @@ private Response process(MultivaluedMap<String, String> params) {
 
         checkSsl();
         checkRealm();
+
+        try {
+            session.clientPolicy().triggerOnEvent(new PreAuthorizationRequestContext(clientId, params));
+        } catch (ClientPolicyException cpe) {
+            return redirectErrorToClient(parsedResponseMode, cpe.getError(), cpe.getErrorDetail());
```

Comment:
`parsedResponseMode` might not be used because it can be set in L158. It might be better to use other error processing method.

Reply:
Thanks for your feedback guys. Let's get this done! :muscle:

As for postponing error handling until redirect URI (and therefore client ID!) is verified - I'm afraid this makes little sense, as the main purpose for this hook is to be able to "legitimize" client IDs that would otherwise be considered invalid by Keycloak. More specifically, the main use case is to automatically register clients that use the "client ID as a URL" paradigm and publish their metadata under that URL. This approach is used not only by Solid-OIDC, but also by [IndieAuth](https://indieauth.net/), [OpenID Connect Federation](https://openid.net/specs/openid-connect-federation-1_0.html) and [OAuth 2.0 Client Discovery](https://datatracker.ietf.org/doc/draft-looker-oauth-client-discovery/01/).

That said, I have changed error handling to display an error page instead of returning a HTTP redirect. This also changes the behavior of the RejectRequestExecutor - when used in the context of Authorization request, it will now fail earlier and display an error page. I have also added a proper integration test for the PreAuthorizationRequest event.